drivers/sdhc/sam_hsmci.c

```diff
@@ -0,0 +1,538 @@
+
+	do {
+		sr = hsmci->HSMCI_SR;
+		if (sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE | HSMCI_SR_DTOE | HSMCI_SR_DCRCE)) {
+			LOG_DBG("PDC sr 0x%08x error", sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE |
+							     HSMCI_SR_DTOE | HSMCI_SR_DCRCE));
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_RXBUFF));
+
+	do {
+		sr = hsmci->HSMCI_SR;
+		if (sr & (HSMCI_SR_UNRE | HSMCI_SR_OVRE | HSMCI_SR_DTOE | HSMCI_SR_DCRCE)) {
+			return -EIO;
+		}
+	} while (!(sr & HSMCI_SR_XFRDONE));
+	return 0;
+}
+
+static int sam_hsmci_write_timeout(Hsmci *hsmci, int timeout_ms)
+{
+	/* convert to clocks (coarsely) */
+	int clocks = ATMEL_SAM_DT_CPU_CLK_FREQ_HZ / 1000 * timeout_ms;
+	int mul, max_clock;
+
+	for (int i = 0; i < _mul_shift_size; i++) {
+		mul = 1 << _mul_shift[i];
+		max_clock = 15 * mul;
+		if (max_clock > clocks) {
+			hsmci->HSMCI_DTOR = ((i << HSMCI_DTOR_DTOMUL_Pos) & HSMCI_DTOR_DTOMUL_Msk) |
+					    HSMCI_DTOR_DTOCYC((clocks + mul - 1) / mul);
+			return 0;
+		}
+	}
+	/*
+	 * So, if it is > maximum timeout... we'll just put it on the maximum the driver supports
+	 * its not nice.. but it should work.. what else is there to do?
+	 */
+	hsmci->HSMCI_DTOR = HSMCI_DTOR_DTOMUL_Msk | HSMCI_DTOR_DTOCYC_Msk;
+	return 0;
+}
+
+static int sam_hsmci_request_inner(const struct device *dev, struct sdhc_command *cmd,
+				   struct sdhc_data *sd_data)
+{
+	const struct sam_hsmci_config *config = dev->config;
+	struct sam_hsmci_data *data = dev->data;
+	Hsmci *hsmci = config->base;
+	uint32_t sr = 0;
+	uint32_t cmdr = 0;
+	uint32_t size;
+	int ret = 0;
+	bool is_write = false;
+
+	LOG_DBG("%s(opcode=%d, arg=%08x, data=%08x, rsptype=%d)", __func__, cmd->opcode, cmd->arg,
+		(uint32_t)sd_data, cmd->response_type & SDHC_NATIVE_RESPONSE_MASK);
+
+	if (cmd->opcode == SD_GO_IDLE_STATE) {
+		/* send 74 clocks, as required by SD spec*/
+		sam_hsmci_send_clocks(hsmci);
+	}
+
+	if (sd_data) {
+		cmdr |= HSMCI_CMDR_TRCMD_START_DATA;
+
+		ret = sam_hsmci_write_timeout(hsmci, cmd->timeout_ms);
+		if (ret != 0) {
+			return ret;
+		}
+
+		switch (cmd->opcode) {
+		case SD_WRITE_SINGLE_BLOCK:
+			cmdr |= HSMCI_CMDR_TRTYP_SINGLE;
+			cmdr |= HSMCI_CMDR_TRDIR_WRITE;
+			is_write = true;
+			break;
+		case SD_WRITE_MULTIPLE_BLOCK:
+			is_write = true;
+			cmdr |= HSMCI_CMDR_TRTYP_MULTIPLE;
+			cmdr |= HSMCI_CMDR_TRDIR_WRITE;
+			break;
+		case SD_APP_SEND_SCR:
+		case SD_SWITCH:
+		case SD_READ_SINGLE_BLOCK:
+			cmdr |= HSMCI_CMDR_TRTYP_SINGLE;
+			cmdr |= HSMCI_CMDR_TRDIR_READ;
+			break;
+		case SD_READ_MULTIPLE_BLOCK:
+			cmdr |= HSMCI_CMDR_TRTYP_MULTIPLE;
+			cmdr |= HSMCI_CMDR_TRDIR_READ;
+			break;
+		case SD_APP_SEND_NUM_WRITTEN_BLK:
+			break;
+		default:
+			return -ENOTSUP;
+		}
+
+		if ((sd_data->block_size & 0x3) == 0 && (((uint32_t)sd_data->data) & 0x3) == 0) {
+			size = (sd_data->block_size + 3) >> 2;
+			hsmci->HSMCI_MR &= ~HSMCI_MR_FBYTE;
+		} else {
+			size = sd_data->block_size;
+			hsmci->HSMCI_MR |= HSMCI_MR_FBYTE;
+		}
+
+		hsmci->HSMCI_MR |= HSMCI_MR_PDCMODE;
```

Comment:
There is no PDC driver at moment. Can you clarify?
I believe this is the only bit that not allow use of this driver with all SoC's.

Reply:
The PDC (Peripheral DMA Controller) is a DMA controller which is combined with the specific peripheral in the same hardware 'IP'. Due to this the SAM4E has PDC registers appended in the memory region of the HSMCI. So, the PDC usage is inside of the driver, like here:

https://github.com/zephyrproject-rtos/zephyr/blob/800cad694002b76ea4016feb04b1f6b678a7887e/drivers/sdhc/sam_hsmci.c#L423-L430

All registers here, except for PTCR, are PDC registers. 

Other parts of AMTEL (I have to look  up which) also have the HSMCI peripheral, but use a generic DMA peripheral to execute the actual DMA transfer. In that case you would need to actually support the DMA controller specific to that part. Which is something I can't test, so will be difficult for me to write.

However, you do have a point, as it should also perfectly possible to not use the PDC, if not available. This will be a bit slower, but should allow support for a more broad range of parts, and I can test this on my own chip. I will see if I can incorporate this into the driver.